### PR TITLE
SDK: Refresh Token implementation

### PR DESCRIPTION
**Pull request by langleyd**
Supports changes in https://github.com/vector-im/element-ios/pull/5293 to resolve https://github.com/vector-im/element-ios/issues/5292

### Included in this PR
- Introduces a new SDK option `MXSDKOptions.authEnableRefreshTokens` to enable refresh tokens. Access token auth will remain in place for the user while logged. Refresh token auth will be used on next login/register.
- Implemented the login/register, refresh endpoints as detailed in https://github.com/vector-im/element-ios/issues/5292.
- The refresh token are now stored on `MXCredential`, it has a function for updating the tokens when they are refreshed. It is the source of truth for these values.
- Previously the access token was long-lived, was passed around and multiple references to it were held without ill-effect. Now that access tokens are short-lived, the approach is to have `MXCredential` be the source of truth and other classes that require it such as `MXHTTPClient` to pull in the value on demand.
- `MXHTTPClient` `requestWithMethod` functions in the past forked, with half of the function signatures supporting an auth mechanism for the Identity server and the others being for homeserver requests. Rather than maintaining two different auth mechanisms for the http client there I created one that servers both identity and homeserver auth. It comes at the risk of updating the identity server auth but with the benefit of approach/code actually being understandable(vs having two implementations).
- `MXHTTPClient` is designated as `authenticated` or not in the init(the later meaning the auth closures are not invoked) and similar to the old identity server auth, individual requests can be downgraded to remove the auth headers. All auth requests are now wrapped in an access token check that verifies its validity/requests a new token if required. All authenticated responses are also checked for the `UNKNOWN_TOKEN` and retried after again verifying/requesting an access token.
- MXRestClients used to interact with the homeserver are responsible for the refresh token exchange. A dispatch group and queue are used to ensure that the first request for an access token that requests an exchange with the server leads and all other requests for the token block until the leader completes. When they are unblocked the new token is in place.
- Refreshes can also take place in other processes like extensions. To ensure there is only 1 leader across processes we do a read + write of the credentials from disk effectively within a single transaction using `NSFileCoordinator`. If the read tells use the credential is good we us it, if not we refresh the token with the server and write it back within the transaction so that the next read get is and we don't have race condition.

### TODO
- [x] Finish preemptive request of tokens at set interval before expiration.
- [x] Make sure with the new code and in place and `MXSDKOptions.authEnableRefreshTokens` enabled that access token auth continues to work until the users logs out.
- [x]  Make sure with the new code and in place and `MXSDKOptions.authEnableRefreshTokens` disabled that access token auth continues to work.

### CODE REVIEW

**Comment by ismailgulek** [comment]
I wonder if it makes sense to have default valued parameters on `MXRestClient` Swift initializer. We had to make a lot of changes in the tests. Totally up to you.

File: MatrixSDK/Data/MXCredentials.m [hunk 1]
```diff
@@ -119,17 +131,21 @@ - (BOOL)isEqual:(id)other
 
     return [_homeServer isEqualToString:otherCredentials.homeServer]
         && [_userId isEqualToString:otherCredentials.userId]
-        && [_accessToken isEqualToString:otherCredentials.accessToken];
+        && [_accessToken isEqualToString:otherCredentials.accessToken]
+        && _accessTokenExpiresAt == otherCredentials.accessTokenExpiresAt
+        && ((_refreshToken == nil && otherCredentials.refreshToken == nil) || [_refreshToken isEqualToString:otherCredentials.refreshToken]);
```
[the remark this answers is not in the document]
Comment (langleyd) [reply MatrixSDK/Data/MXCredentials.m:136]: Sorry I do need this check. In cases where both credentials have a nil refresh token we want them to be assumed equal. But if they are both nil the righthand side of the or will return NO.

[the remark this answers is not in the document]
Comment (ismailgulek) [reply MatrixSDK/Data/MXCredentials.m:136]: Yeah you're right. Sorry for the wrong direction.

File: MatrixSDK/JSONModels/MXRefreshResponse.h [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#import "MXJSONModel.h"
+
+/**
+ `MXRefreshResponse` represents the response to an auth refresh request.
+ */
+@interface MXRefreshResponse : MXJSONModel
```
Comment (ismailgulek) [inline MatrixSDK/JSONModels/MXRefreshResponse.h:22]: Please use nullability attributes for this class too. Sorry if i haven't pointed out that before.

**Pull request merged**